src/controllers/legacycontrollermapping.cpp

```diff
@@ -48,14 +56,36 @@ void LegacyControllerMapping::loadSettings(const QFileInfo& mappingFile,
     }
 }
 
+void LegacyControllerMapping::resetSettings(const QFileInfo& mappingFile,
+        UserSettingsPointer pConfig,
+        const QString& controllerName) {
+    QString controllerPath =
+            mappingFile.absoluteFilePath()
+                    .replace(pConfig->getSettingsPath(),
+                            kControllerSettingsSettingPathSubst)
+                    .replace(pConfig->getResourcePath(),
+                            kControllerSettingsResourcePathSubst);
+
+    QString controllerKey = QString(kControllerSettingsPreferenceGroupKey)
+                                    .arg(controllerName, controllerPath);
+
+    QList<ConfigKey> definedSettings = pConfig->getKeysWithGroup(controllerKey);
+    for (const auto& key : definedSettings) {
+        VERIFY_OR_DEBUG_ASSERT(pConfig->remove(key)) {
+            qDebug() << "Unable to delete a controller" << key
+                     << "- As the setting been removed already?";
+        }
```

Comment:
Ah yes, you're right.

I have refactored slightly the flow and cleanup some of the convoluted load/restore.

I believe it works as expected now.

Reply:
Unfortunately, it still does not work correct. While the `Cancel`, `Ok` and `Apply` buttons work now as expected, the `Restore Defaults` does not anymore. It doesn't restore the default values specified in the XML file - it now reverts the changes made since opening the preferences dialog.

Reply:
Currently, you need to `apply` or `ok` after `Restore Defaults`. If you `Restore default` then select again the mapping, you undo the restore as per [the comment](https://github.com/mixxxdj/mixxx/blob/bf7830de5ae21eacbc8e815a4c84884ba7ae1e2b/src/controllers/dlgprefcontroller.cpp#L861) here.

https://github.com/mixxxdj/mixxx/assets/7086688/a97722bc-4c82-427f-97f6-c2d18aeadf10

I could put a condition in `slotShowMapping` which doesn't load settings if the current selected mapping is the same than the loaded one but that sounds hacky. otherwise, I have no idea how to fix that without significant more change.

Reply:
No, leave this as it is, this is all fine.

To reproduce the issue, do the following steps:

1. Opening the settings dialog of a mapping
2. Modify a value manually and close the dialog
3. Open the settings dialog again
4. Modify the value again
5. Press the `Restore Defaults` button
-> Expected is that the GUI shows the default value defined in the XML file, but it shows the value set in step 2. instead

Reply:
I assume you've selected the mapping again after step `5`, right?
This discards unsaved changes, including the reset, so it brings the value back to the currently saved one, in step `2`. 

I have tried your scenario with a `ok` or `apply` after the `restore default` and the GUI shows the default value defined in the XML file after that.

https://github.com/mixxxdj/mixxx/assets/7086688/5a4a6788-6dbb-4e56-bf6c-bd1a4f05d353

Reply:
No, it doesn't unload the mapping for me (and it should not):


https://github.com/mixxxdj/mixxx/assets/64457745/a876ff0d-04db-458c-abe8-3e6a54c66b91

Reply:
Oh right, looks like we got a bug on Linux.

I'll make some changes so it redraws the settings in the restore slot

Reply:
Okay, it took my quite some time and added quite some complexity unfortunately. Hopefully it know work as expected. I'm afraid clazy will fail, but I will wait till you confirm it's working before looking into it

Reply:
It works now for the case above, but not for the much simpler case, that I open a mapping with default values, edit them and press "Restore Defaults" before saving the settings by Ok or Apply.

Reply:
The last approach looks far too complicated to me. I suggest to leave the mapping unload out here, since I guess this also occurs without this PR.
Since there is already already function, that compares edited value with default value, it can't be too difficult to create one that sets edited values to default values.
Than there is only a function needed, to set the widgets to the edited values - this must not consider, which controller/mapping we have, because the edited value will always be the truth.

Reply:
> The last approach looks far too complicated to me. I suggest to leave the mapping unload out here, since I guess this also occurs without this PR. 

Agreed, will revert

> Since there is already already function, that compares edited value with default value, it can't be too difficult to create one that sets edited values to default values. Than there is only a function needed, to set the widgets to the edited values - this must not consider, which controller/mapping we have, because the edited value will always be the truth.

The problem is - when you click reset, a new mapping is reloaded, even if the selected value is the same; it recreates the same one. This means that the setting layout gets redrawn since a new mapping is detected (thus why trying to implement mapping comparison on the current commit), so the widgets get recreated. That means that the dirty value is lost, except if it gets saved just before, which we don't want to do till apply. 

I'm going to try a new approach tomorrow using `find` and dynamic cast, except if you see something else than can work with the existing state of things?

Reply:
Right, so I believe I managed to fix that issue with reset and also manage to clean up slightly those unneeded UI updates... :crossed_fingers:

Reply:
Unfortunately, the new behavior is also not correct. See the video, where I clicked multiple times on the `Restore Defaults` button:

https://github.com/mixxxdj/mixxx/assets/64457745/58ec75e8-5bef-412e-bf88-334e14d69630

Why did you replace make_parented by new at several places? I don't found this memory released anywhere later:
![grafik](https://github.com/mixxxdj/mixxx/assets/64457745/f5fc18e8-483b-49c6-b6ed-cb8caa3ecc17)

Reply:
> Why did you replace make_parented by new at several places?

Because I need to capture the object pointer in the anonymous slot right under this allocation, and `parent_ptr` doesn't implement the copy constructor. Arguably, I could have created a templated class to wrap the QWidget and take care of the signal without anonymous slot, but that's a significant amount of work.

> I don't found this memory released anywhere later

The memory is released here:
```
qDeleteAll(m_ui.groupBoxSettings->findChildren<QWidget*>("", Qt::FindDirectChildrenOnly));
```
Also note these widgets are all parented so them will naturally be deleted when the parent is as per [the doc](https://doc.qt.io/qt-6/qobject.html#details)

> it will automatically delete its children in its destructor

I've pushed attempt#687 :roll_eyes: